### PR TITLE
Fix for smart_decorator

### PR DESCRIPTION
**Pull request by MegaIng**
- [x] new test to catch edge cases.
- [x] find all edge cases? (@erezsh and @ornariece I would like to know if you find more edge cases)
- [x] rework `smart_decorator`

### CODE REVIEW

**Comment by MegaIng** [comment]
Ok, this is pretty much finished. The only missing edge case is when you use a staticmethod from one class in another. That is as far as I can tell impossible to differentiate from just a normal function definition. I think this is a complete alternative to #999, and should fix all problems encountered in that area. (Except pickling, but that is a different PR).

File: tests/test_trees.py [hunk 1]
```diff
@@ -171,6 +173,76 @@ def integer2(cls, args):
...
+            # "oc_class_ab_classmethod": ([1, 2], (1, 2)),
+            # "oc_class_ab_staticmethod": ([1, 2], (1, 2)),
+            # "oc_ab_method": ([1, 2], (1, 2)),
+            # "custom_callable": ([1, 2], (1, 2)),
+            "set_union": ([["b"], ["c"]], {"a", "b", "c"}),
+            "static_add": ([1, 2], 3),
+            "partial_reduce_mul": ([[1, 2]], 2),
+        }
+        for method_name, (children, expected_result) in expected.items():
+            result = test_instance.transform(Tree(method_name, children))
```
Comment (ornariece) [inline tests/test_trees.py:253]: shouldn't we aslo test the static/class methods with the uninstanciated class? you can have a perfectly functional resulting method but that didn't preserve the staticity

Comment (MegaIng) [reply tests/test_trees.py:253]: Yes, will add that.

**Comment by MegaIng** [comment]
Can you provide a small example for that?

**Comment by ornariece** [comment]
```python
@v_args(inline=True)
class T(Transformer):
    @staticmethod
    def test():
        pass


@v_args(inline=True)
class T2(Transformer):
    test = T.test


T2.test()
```

**Comment by MegaIng** [comment]
That edge case still exist and is the same one as for non-v_args function, and also the expected behavior from a python standpoint. We can however special case and make it work...

```
class T:
    @staticmethod
    def test():
        pass


class T2:
    test = T.test


T2().test()
```

also doesn't work.

**Comment by ornariece** [comment]
oh right yes, it's the exact same as using a normal function, for which you'd have to redeclare the staticity. i'd keep it that way, as it's what you'd expect from classic python classes

**Comment by MegaIng** [comment]
Yes, It was also a TIL for me, but I wouldn't deviate in that respect from python. The current implementation should now always exactly reflect what python would be doing.

**Comment by ornariece** [comment]
i've run my project (which makes an extensive use of ``v_args`` decorators) with your branch and every output was correct 👍

**Comment by ornariece** [comment]
@erezsh could you review this? it should be ready for merging, unless new edgecases are found

File: lark/visitors.py [hunk 2]
```diff
@@ -402,46 +401,47 @@ def inner(cls, tree):
...
-    f.vargs_applied = True
-    f.visit_wrapper = visit_wrapper
-    return f
+    def __set_name__(self, owner, name):
+        try:
+            f = self.base_func.__set_name__
+        except AttributeError:
+            return
+        else:
+            f(owner, name)
```
Comment (erezsh) [inline lark/visitors.py:444]: This isn't covered by any of the tests

**Comment by erezsh** [comment]
Interesting approach! Definitely seems better.

**Comment by erezsh** [comment]
Also, `examples\advanced\create_ast.py` is failing for this branch.

File: lark/visitors.py (hunk 2 above)
Comment (MegaIng) [reply lark/visitors.py:444]: That is because I don't know when it might be relevant. AFAIK no standard descriptors (except maybe `property`) actually use `__set_name__`. I just added this for completeness. But I will see if I find something.

**Comment by MegaIng** [comment]
What is this error message: `cannot convert dictionary update sequence element #0 to a sequence`? WTF does that mean?

(And `create_ast` was a very simple fix.)

> Definitely seems better

Also, ignoring the tests this is one of those rare PRs that remove more code that they add :-).

**Comment by MegaIng** [comment]
The problem in `dynamic_complete.py` was not caused by this PR, but by the change of the default for `maybe_placeholders`.

File: lark/visitors.py (hunk 2 above)
Comment (MegaIng) [reply lark/visitors.py:444]: I added a test. Not the most useful of situations, but one that breaks without the above code (and never would have worked with the old system).

**Comment by erezsh** [comment]
The test looks okay. Where's the create_ast fix? :)

File: lark/visitors.py [hunk 3]
```diff
@@ -427,7 +427,7 @@ def __init__(self, func: Callable, visit_wrapper: Callable[[Callable, str, list,
     def __call__(self, *args, **kwargs):
         return self.base_func(*args, **kwargs)
 
-    def __get__(self, instance, owner):
+    def __get__(self, instance, owner=None):
```
Comment (MegaIng) [inline lark/visitors.py:430]: This is the `create_ast.py` fix.

**Comment by erezsh** [comment]
Always happy to remove code.
